sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp

```diff
@@ -135,6 +137,45 @@ namespace Azure { namespace Core { namespace Http { namespace Policies {
    */
   struct TransportOptions final
   {
+#if !defined(BUILD_TRANSPORT_CUSTOM_ADAPTER)
```

Comment:
?? We have many cases where we have ifdef conditionals like this in headers. See also: context.hpp, dll_import_export.hpp, http.hpp, win_http_transport.hpp, body_stream.hpp, etc.

Reply:
dll_import_export does rely on install to have that file patched.
win_http_transport may get included by customer code, when they build with winhttp support.
http.hpp has ifNdef(build_testing)
body_strema has ifdefs for AZ_PLATFORM which WILL work.

But of you are including policy.hpp from a customer code, the customer will not have `BUILD_TRANSPORT_CUSTOM_ADAPTER` defined. That macro is safe to use only in the SDK cpp files.

Reply:
`BUILD_*` is only defined when the Azure SDK is getting built, not when the customer code consumes it.

Reply:
What is your suggestion then: 

We only want these options to be available when the customer is building the SDK and configuring WinHTTP or CURL. 

How do I ensure that these options are only present when the customer is building either the WinHTTP or CURL transports?

Reply:
1. They can be present at all times, regardless of transport
2. Maybe we can move them to the corresponding headers for each transport

Reply:
We had a long discussion about that in email (you might have been OOF during that discussion) and the consensus on the team was that their presence would be confusing to developers who provided their own `Transport`. Honestly, I'd *love* to just make them global and remove the whole conditional thingy, but the consensus on the team was that their presence would be confusing. Personally, I don't completely understand that concern - I think that the rules are pretty straightforward, but there was a clear concern among the team members about how to present these options which will apply to any transport.

There are analogs to each of these in the corresponding headers. The point of these fields is that these are options that are (a) commonly available across multiple HTTP transports and (b) don't require a specific HTTP transport to be chosen.

Reply:
Btw, as an example of how this could be use: If we want to use the TestProxy, currently we need to manually configure a Transport object in the `TestBase` class. With this change, we just set the `HttpProxy` and let the TransportPolicy choose an appropriate transport. So it's a really useful feature.

Reply:
Also: The transport specific analogs of these fields are each tied to the individual transport's semantics. So for CURL, `HttpProxy` takes a std::string containing a set of PEM encoded X.509 certificates. For WinHTTP, it takes a std::vector<std::string> each of which is a Base64 encoded X.509 certificate. That's because WinHTTP takes a different set of inputs than CURL. The common options smooth out the differences between the transports.

Reply:
how is it expected to work?
We have a header file:
```cpp
#ifdef (MACRO)
void f();
#endif
```

We compile our SDK: `cmake -DMACRO`. Everything does compile, because the macro is set.

User installs the package: `vcpkg install azure-sdk`

User uses the SDK:
```cpp
#include <azure/header.hpp>

int main()
{
    f(); // <-- not defined, because MACRO is not set
}
```

Reply:
We have been avoiding relying on these macros up until now.

Reply:
There are many cases in our headers where we rely on these (or similar) macros. One easy example is `BUILD_TRANSPORT_CUSTOM_ADAPTER`. 

If we can't depend on macros, how do we achieve the consensus goal: We don't want these values to show up in the headers if the customer has not specified that they want to use either WinHTTP or CURL?

I'm completely open to any other suggestions about how to implement the feature request of "We only want these values to show up in the headers if the customer is using either WinHTTP or CURL". If my idea of using #defines to achieve this doesn't work, please let me know how to achieve the goal without them.

Reply:
we don't rely on them in our public headers. Only in sources, because they are effective only when we compile the SDK.
You won't find an opposite example.

Solution is to move options to the corresponding transport header, or to make these options available regardless of transport, without ifdef.

Otherwise, customers will need to define `BUILD_CURL_CUSTOM_ADAPTER` when building THEIR code. That's not right.

Reply:
Plus, it should be possible to build with both transport adapters, and instantiate one client with WinHTTP, another with libcurl transport, third with the custom transport. @JeffreyRichter wanted for this scenario to be possible. What the macro should be then? Effectively, options are going to be visible for every transport adapter, even if the custom one does not support some of them. Plus, if the cutomer has `BUILD_CURL_HTTP_TRANSPORT_ADAPTER` defined in one translation unit, but not in the other, the code will compile, but the options are going to be binary incompatible, and the SDK they have built with all the transports, is going to read garbage from the memory, because the customer did not have `BUILD_CURL_HTTP_TRANSPORT_ADAPTER` defined in THEIR project.

Reply:
The macro is "if building for WInHTTP or building for CURL" If you're building for both WinHTTP and CURL, that's just fine ( `(1 || 1) == 1`).

I'm open to removing this conditional, but Jeff and Ahson raised significant objections to having global options. And I feel very strongly that we SHOULD have global options.

Reply:
We can have global options (nothing wrong with that), we can have per-transport options (also not bad), but we can't have ifdef in the public header that checks for a variable that is only defined when the SDK is being built, and not when it is being consumed.

About "`(1 || 1)`", I get that. But if in different translation units customer have it defined/not defined, then the code is going to compile, but the struct will not be binary compatible with the SDK that was built WITH the transport support.
And I don't like asking the customer to define `BUILD_CURL_HTTP_TRANSPORT_ADAPTER` (right, it does not have `AZURE_SDK_` prefix) when building THEIR code. And that requirement is a breaking change.

Currently, I build SDK with transport support, I keep consuming the code as before, without defining anything, and what struct does my code instantiate? It instantiates empty options struct. The struct gets passed to the SDK code, and what it thinks the struct should look like? It thinks it has all the fields. And "reads" them. From a random memory.

Reply:
If customers are building with different settings for different translation units, they are 100% going to be hitting ODR violations (literally - ODR violations occur when you have different objects built with different definitions).

IMHO we should never design with the assumption that customers are intentionally introducing ODR violations - therein lies madness.

Having said all that, I'll pull out the transport specific macros (@ahsonkhan and @JeffreyRichter can chime in if they want them back).

I'm going to keep the `!BUILD_TRANSPORT_CUSTOM_ADAPTER` because that is a build setting that customers set (we never set BUILD_TRANSPORT_CUSTOM_ADAPTER). Note that without this macro being defined, the customers `AzSdkGetCustomHttpTransport` method will never be called, so if the customer is depending on `AzSdkGetCustomHttpTransport`, they will have to define this in their build configuration.

Reply:
Resolved: The core issue is that the transport adapter options aren't available to customers.